src/main/java/net/wimpi/modbus/io/ModbusTCPRTUTransport.java

```diff
@@ -0,0 +1,218 @@
+        throw new RuntimeException("Operation not supported.");
+    } // readRequest
+
+    @Override
+    public ModbusResponse readResponse() throws ModbusIOException {
+        ModbusResponse response = null;
+        int dlength = 0;
+
+        try {
+            // read to function code, create request and read function specific bytes
+            synchronized (m_ByteIn) {
+                int uid = m_Input.read();
+                logger.trace("Managed to read at least one byte");
+                if (uid != -1) {
+                    int fc = m_Input.read();
+                    m_ByteInOut.reset();
+                    m_ByteInOut.writeByte(uid);
+                    m_ByteInOut.writeByte(fc);
+
+                    // create response to acquire length of message
+                    response = ModbusResponse.createModbusResponse(fc);
+                    response.setHeadless();
+
+                    getResponse(fc, m_ByteInOut);
+
+                    dlength = m_ByteInOut.size() - 2; // less the crc
+                    logger.debug("Response: {}", ModbusUtil.toHex(m_ByteInOut.getBuffer(), 0, dlength + 2));
+
+                    m_ByteIn.reset(m_InBuffer, dlength);
+
+                    // check CRC
+                    int[] crc = ModbusUtil.calculateCRC(m_InBuffer, 0, dlength); // does not include CRC
+                    if (ModbusUtil.unsignedByteToInt(m_InBuffer[dlength]) != crc[0]
+                            || ModbusUtil.unsignedByteToInt(m_InBuffer[dlength + 1]) != crc[1]) {
+                        throw new IOException("CRC Error in received frame: " + dlength + " bytes: "
+                                + ModbusUtil.toHex(m_ByteIn.getBuffer(), 0, dlength));
+                    }
+                } else {
+                    throw new IOException("Error reading response (EOF)");
+                }
+
+                // read response
+                m_ByteIn.reset(m_InBuffer, dlength);
+                response.readFrom(m_ByteIn);
+            } // synchronized
+            return response;
+        } catch (Exception ex) {
+            final String errMsg = "failed to read";
+            logger.debug("Last request: {}", ModbusUtil.toHex(lastRequest));
+            logger.debug("{}: {}", errMsg, ex.getMessage());
+            throw new ModbusIOException(
+                    String.format("I/O exception: %s %s", ex.getClass().getSimpleName(), ex.getMessage()));
+        }
+    }// readResponse
+
+    @Override
+    protected void prepareStreams(Socket socket) throws IOException {
+        m_Input = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
+        m_Output = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
+        m_ByteOut = new BytesOutputStream(Modbus.MAX_MESSAGE_LENGTH);
+        m_InBuffer = new byte[Modbus.MAX_MESSAGE_LENGTH];
+        m_ByteIn = new BytesInputStream(m_InBuffer);
+        m_ByteInOut = new BytesOutputStream(m_InBuffer);
```

Comment:
call super.prepareStreams and init only the new fields here

Reply:
Not really. Two of the streams declared in super can be created in the super class, but The third one is created in the inherited class using a **_dfferent_** constructor. So rather than creating it in super, throwing it away, and recreating in inherited, I think it is best if all contractors are called in the class where they belong.

Reply:
Makes sense, did not notice this.